<a href="https://colab.research.google.com/github/Alfredosernas/Machine-Learning/blob/main/Hands_on_5_knn.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Clasificaccion KNN
 K-Nearest Neighbors (KNN) es un modelo de aprendizaje automatico supervisado.KNN hace predicciones basadas en lo similares que son las observaciones de entrenamien a las nuevas observaciones entrantes y sin etiquetar.

 EL número de vecionos más cercanos .o "K", es el principal factor decisivo en KNN.El principio de voto mayoritario esta en el nucleo del aprendizaje de KNN.


#Funcionamiento del algoritmo KNN
1. El algoritmo KNN calcula la distancia desde los datos de prueba hasta cada fila de datos de entrenamiento.
2.La distancia puede calcularse usando la distancia euclidiana, de Manhattan,de Hamming o de Minkowski.generalmente , se utiliza la distancia euclidiana.
3. La distancia se ordena entonces en orden ascendente,de donde se eligen los K valores superiores.
4. En la mayoria de las clases, la clasificacion se realiza con datos de prueba.


#Implementacion
## Paso:1 Recopilacion de datos
El primer paso es leer los datos que usaremos como entrada.Para este,utilizaremos el conjunto de datos del titanic. para la carga de datos utilizamos la libreria pandas.

In [6]:
#instalamos la librerias y el dataset
import numpy as np
import pandas as pd
df=pd.read_csv('/content/titanic (1).csv')
df.head()

,PassengerId,Pclass,Name,Sex,Age,Ticket,Fare,Survived
0,1,3,"Braund, Mr. Owen Harris",male,22.0,A/5 21171,7.2500,0
1,2,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,PC 17599,71.2833,1
2,3,3,"Heikkinen, Miss. Laina",female,26.0,STON/O2. 3101282,7.9250,1
3,4,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,113803,53.1000,1
4,5,3,"Allen, Mr. William Henry",male,35.0,373450,8.0500,0


## paso 2: Preprocesamiento de datos

### relleno de valores faltantes
comoprobamos el numero de valores que faltan para cada caracteristica usando la funcion "isnull".cuando llamamos a la funcion "suma" junto con isnull, la suma de los datos faltantes en cada columna es la salida.




In [7]:
#Finding the number of missing values for each feature
df.isnull().sum()

,0
PassengerId,0
Pclass,0
Name,0
Sex,0
Age,177
Ticket,0
Fare,0
Survived,0


Podemos observar que Edad de Columna tiene un total de 177 datos faltantes. Como esta columna es numérica, podemos rellenar los valores que faltan con valores estadísticos como media, mediana o modo. En este caso, utilizaríamos la media de los valores no nulos para rellenar los datos faltantes.

In [11]:
#finding and replacing the missing values with mean of non-null values of age
df['Age']=df['Age'].replace(np.nan,df['Age'].mean())
df['Age'][:10]

,Age
0,22.000000
1,38.000000
2,26.000000
3,35.000000
4,35.000000
5,29.699118
6,54.000000
7,2.000000
8,27.000000
9,14.000000


##Codificación de datos categóricos
Luego usamos una estrategia básica de "buscar y reemplazar" para codificar los datos categóricos. "Masculino" y "femenino" son los valores categóricos en la columna "Sexo". El macho se reemplaza por 0 mientras que el hembra se reemplaza por 1.

In [13]:
#encoding categorical data to numerical:
Gender = {'Sex':{'male': 0, 'female': 1}}
df = df.replace(Gender)
df.head()

/tmp/ipykernel_2223/3372115038.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df = df.replace(Gender)


,PassengerId,Pclass,Name,Sex,Age,Ticket,Fare,Survived
0,1,3,"Braund, Mr. Owen Harris",0,22.0,A/5 21171,7.2500,0
1,2,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",1,38.0,PC 17599,71.2833,1
2,3,3,"Heikkinen, Miss. Laina",1,26.0,STON/O2. 3101282,7.9250,1
3,4,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",1,35.0,113803,53.1000,1
4,5,3,"Allen, Mr. William Henry",0,35.0,373450,8.0500,0


##Eliminación de datos irrelevantes
Las características irrelevantes en los datos pueden disminuir la precisión del modelo y hacer que aprenda basándose en características irrelevantes. Estas características pueden eliminarse automáticamente o manualmente.
Características como nombre, número de pasajero, billete, no proporcionan detalles adicionales sobre las variables objetivo. Por lo tanto, eliminamos estas columnas usando funciones de drop


In [15]:
df.drop(['Name', 'PassengerId', 'Ticket'], axis=1, inplace=True)
df.head()

,Pclass,Sex,Age,Fare,Survived
0,3,0,22.0,7.2500,0
1,1,1,38.0,71.2833,1
2,3,1,26.0,7.9250,1
3,1,1,35.0,53.1000,1
4,3,0,35.0,8.0500,0


#Paso 3: Decide qué modelo.
Como la variable objetivo es categórica, realizaríamos una tarea de clasificación con el algoritmo KNN

#Paso 4: Divide el conjunto de datos.
Primero, dividiremos el conjunto de datos en entradas y objetivo. La entrada será en todas las columnas excepto en 'Sobrevivido' porque será la variable objetivo. Para ello, utilizaremos la función de "drop" de Panda.

In [16]:
# Split the dataset into input and target features
X=df.drop('Survived', axis=1)
y=df.Survived

El conjunto de datos se dividirá entonces en datos de entrenamiento y prueba. Los datos de entrenamiento se usarán para entrenar el modelo, mientras que los datos de prueba se usarán para ver qué tan bien funciona el modelo sobre datos que no se han visto antes
La función "train_test_split" en Scikit-Learn puede usarse para separar nuestro conjunto de datos.

In [17]:
# Split the input and target features into training and testing set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=1)

El parámetro test_size se ha establecido en 0,3. Esto significa que el 30% de los datos se utilizará para pruebas, mientras que el 70% restante se utilizará para fines de entrenamiento. Poner un random_state a 1 garantiza que la división será la misma cada vez.

#Paso 5: Entrena el modelo.
El siguiente paso es entrenar el modelo.

In [18]:
from sklearn.neighbors import KNeighborsClassifier
# create knn classifier
knn = KNeighborsClassifier(n_neighbors = 3)
#fit the data
knn.fit(X_train,y_train)

KNeighborsClassifier(n_neighbors=3)

Utilizamos la biblioteca integrada sklearn KNeighborsClassifier para este propósito. "n_neighbors" se ha puesto en 3. La mayoría de los tres vecinos más cercanos se utilizará para etiquetar un nuevo punto de datos.
A continuación, usaremos la función "ajustar" para entrenar el modelo, que tiene dos parámetros: X_train y y_train.

#Paso 6: Evaluación
El modelo está ahora listo para generar predicciones basadas en los datos de prueba. Usamos la función de "predecir" para hacer predicciones.

In [19]:
#make prediction
y_pred=knn.predict(X_test)

Finalmente, se utiliza el accuracy_score de la clase de métricas para calcular la precisión.

In [23]:
from sklearn.metrics import accuracy_score,confusion_matrix
knn_accuracy = accuracy_score(y_test, y_pred)
print('Accuracy:', round(knn_accuracy * 100, 2), '%')

Accuracy: 70.15 %


Una matriz de confusión podría utilizarse para entender eficientemente el rendimiento de un modelo de clasificación. Demuestra cómo nuestro modelo se confunde al hacer predicciones.

In [24]:
#Confusion matrix
print(confusion_matrix(y_test,y_pred))

[[125  28]
 [ 52  63]]


Podemos ver los valores de TP = 125, FP = 28, FN = 52, TN = 63

Aquí TP = Verdadero Positivo, FP = Falso Positivo, FN = Falso Negativo, TN = Verdadero Negativo.